src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -242,6 +244,140 @@ protected static boolean is7Bit(CodeRange codeRange) {
 
     }
 
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "encoding"),
+            @NodeChild(type = RubyNode.class, value = "bytes")
+    })
+    @ImportStatic(RopeGuards.class)
+    public abstract static class CalculateAttributesNode extends RubyNode {
+
+        public static CalculateAttributesNode create() {
+            return RopeNodesFactory.CalculateAttributesNodeGen.create(null, null);
+        }
+
+        abstract StringAttributes executeCalculateAttributes(Encoding encoding, byte[] bytes);
+
+        @Specialization(guards = "isEmpty(bytes)")
+        StringAttributes calculateAttributesEmpty(Encoding encoding, byte[] bytes,
+                                                  @Cached("createBinaryProfile()") ConditionProfile isAsciiCompatible) {
+            return new StringAttributes(0,
+                    isAsciiCompatible.profile(encoding.isAsciiCompatible()) ? CR_7BIT : CR_VALID);
+        }
+
+        @Specialization(guards = { "!isEmpty(bytes)", "isBinaryString(encoding)" })
+        StringAttributes calculateAttributesBinaryString(Encoding encoding, byte[] bytes,
+                                                         @Cached("create()") BranchProfile nonAsciiStringProfile) {
+            CodeRange codeRange = CR_7BIT;
+
+            for (int i = 0; i < bytes.length; i++) {
+                if (bytes[i] < 0) {
+                    nonAsciiStringProfile.enter();
+                    codeRange = CR_VALID;
+                    break;
+                }
+            }
+
+            return new StringAttributes(bytes.length, codeRange);
+        }
+
+        @Specialization(rewriteOn = NonAsciiCharException.class,
+                guards = { "!isEmpty(bytes)", "!isBinaryString(encoding)", "isAsciiCompatible(encoding)" })
+        StringAttributes calculateAttributesAsciiCompatible(Encoding encoding, byte[] bytes) throws NonAsciiCharException {
+            // Optimistically assume this string consists only of ASCII characters. If a non-ASCII character is found,
+            // fail over to a more generalized search.
+            for (int i = 0; i < bytes.length; i++) {
+                if (bytes[i] < 0) {
+                    throw new NonAsciiCharException();
```

Comment:
This means we'll walk the 7-bit bytes a second time for potentially every instance of this node.
Maybe the `Specialization` could be made a normal method and take advantage that we know bytes `0...i` are 7-bit?
That would avoid the exception/rewriteOn entirely.
But this is not too bad either, I'm just not too fond of rewriteOn as it seems fairly expensive in interpreter and first executions.

Reply:
This node is basically the byte scan node to determine code range in the absence of any information. What I've done here is extracted the logic from making leaf ropes to a common location where it can be used for both leaf rope creation and substring operation.

I can be perhaps be convinced that it'd be costly for substrings. But it produces very nice, short code when creating leaf ropes in an application that uses 7-bit strings predominately.

Reply:
Let's keep it for now.